src/cloudant/result.py

```diff
@@ -267,3 +269,138 @@ def __iter__(self):
+    Provides a sliceable and iterable interface to query result collections.
+    A QueryResult object is instantiated with the Query
+    :func:`~cloudant.query.Query.__call__` callable, which is used to retrieve
+    data.  A QueryResult object can also use optional extra arguments for result
+    customization and supports efficient, paged iteration over the result
+    collection to avoid large result data from adversely affecting memory.
+
+    In Python, slicing returns by value, whereas iteration will yield
+    elements of the sequence.  This means that slicing will perform better
+    for smaller data collections, whereas iteration will be more
+    efficient for larger data collections.
+
+    For example:
+
+    .. code-block:: python
+
+        # Slicing by value:
+        query_result[100:200] # results between the 100th and the 200th result
+        query_result[:200]    # results up to the 200th result
+        query_result[100:]    # results after 100th result
+        query_result[:]       # all results
+
+        # Iteration:
+
+        # Iterate over the entire result collection
+        query_result = QueryResult(query)
+        for doc in query_result:
+            print doc
+
+        # Iterate over the result collection, with an overriding query sort
+        query_result = QueryResult(query, sort=[{'name': 'desc'}])
+        for doc in query_result:
+            print doc
+
+        # Iterate over the entire result collection,
+        # explicitly setting the index and in batches of a 1000.
+        query_result = QueryResult(query, use_index='my_index', page_size=1000)
+        for doc in query_result:
+            print doc
+    """
+    def __init__(self, query, **options):
+        self.options = options
+        self._query = query
+        self._page_size = options.pop('page_size', 100)
+
+    def __getitem__(self, key):
+        """
+        Provides query result slicing by element support.  Key access and
+        slicing by non-integer key value are not available for query results.
+
+        See :class:`~cloudant.result.QueryResult` for slicing examples.
+
+        :param key:  Must be a range defined by two integers.  Slices
+            will be interpreted as ``skip:limit-skip`` style pairs.
+            For example ``[100:200]`` means skip 100 records
+            then get up to and including the 200th record so that you get the
+            range between the supplied slice values.  Whereas ``[:100]`` means
+            get up to and including the 100th record.
+
+        :returns: Rows data in JSON format
+        """
+        if 'skip' in self.options or 'skip' in self._query.keys():
+            msg = 'Cannot use skip parameter with QueryResult slicing.'
+            raise CloudantArgumentError(msg)
+        if 'limit' in self.options or 'limit' in self._query.keys():
+            msg = 'Cannot use limit parameter with QueryResult slicing.'
+            raise CloudantArgumentError(msg)
+        if (
```

Comment:
I've refactored the QueryResult to extend the Result to minimize code replication.  Done in 1f60cd9.